### PR TITLE
api: add stats_matcher for envoy's stats

### PR DESCRIPTION
**Pull request by tmsnan**
api: add stats_matcher  for envoy's stats
Relates to  #1669

### CODE REVIEW

**Comment by zirain** [comment]
please add api first

**Comment by tmsnan** [comment]
> please add api first

done

File: api/config/v1alpha1/metric_types.go [hunk 1]
```diff
@@ -10,6 +10,13 @@ type ProxyMetrics struct {
 	Prometheus *PrometheusProvider `json:"prometheus,omitempty"`
 	// Sinks defines the metric sinks where metrics are sent to.
 	Sinks []MetricSink `json:"sinks,omitempty"`
+	// ProxyStatMatcher defines configuration for reporting custom Envoy stats.
+	// To redurece memory and CPU overhead from Envoy stats system.
```
Comment (arkodg) [inline api/config/v1alpha1/metric_types.go:14]:
```suggestion
	// To reduce memory and CPU overhead from Envoy stats system.
```

File: api/config/v1alpha1/metric_types.go [hunk 2]
```diff
@@ -10,6 +10,13 @@ type ProxyMetrics struct {
 	Prometheus *PrometheusProvider `json:"prometheus,omitempty"`
 	// Sinks defines the metric sinks where metrics are sent to.
 	Sinks []MetricSink `json:"sinks,omitempty"`
+	// ProxyStatMatcher defines configuration for reporting custom Envoy stats.
```
Comment (arkodg) [inline api/config/v1alpha1/metric_types.go:13]: * what is the default behavior,  is it all stats ? if so can we add a note around that ?
* what happens when the user specifies a match string that doesnt match any metric ?
* should we combine `ProxyStatMatcher` & `HistogramBucketSettings` into one setting ?
    * for a specific `match` of a stat, use this bucket, if bucket unspecified, use default bucket setting 
 * `proxyStatsMatcher` seems too wordy, thoughts on `matches` or `statsSelectors`    
 * can we use the existing style of `Type` used in many places in this project and upstream for `InclusionPrefixes` etc e.g.
 ```
 matches:
 - type: "Prefix"
    value: "cluster_upstream_rq"
 ```

Comment (tmsnan) [reply api/config/v1alpha1/metric_types.go:13]: - The current default behavior is to output metrics prefixed with `cluster_manager, listener_manager, server, cluster.xds-grpc`, I think these are the most basic metrics. Default behaviors need to be discussed，releated  https://github.com/envoyproxy/envoy/issues/28179
- The current metric generation behavior is to only generate metrics matching default rules and matching custom rules. If no custom rules are matched, only the default behavioral metrics are generated.
- ProxyStatMatcher & HistogramBucketSettings can be used to reduce memory and CPU overhead from envoy stats system. Maybe combine them into `customStats` field.
-  For [StatsMatcher](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/metrics/v3/stats.proto#config-metrics-v3-statsmatcher) ，we actually only used the inclusion_list (No stats are enabled except for those matching one of the supplied StringMatcher). It may be clearer to use InclusionXX here

Comment (arkodg) [reply api/config/v1alpha1/metric_types.go]: @AliceProxy @kflynn and I discussed this in the meeting today are here are some thoughts
* lets create a separate issue for Histogram buckets and better understand the type of buckets users wants to configure
* the current API is not very intuitive
```
metrics:
  customMetricScale:
    statsMatcher:
      inclusionMatches:
      - type:  ...
        value: ...

```

there seem be multiple layers of `match` which feels cumbersome

Instead can we stick to the earlier suggestion
```
metrics:
  matches:
  - type: ...
     value: ...
```

and lets document which metrics/stats are computed when
* when `matches` is `nil`
* when `matches` is not empty

Comment (tmsnan) [reply api/config/v1alpha1/metric_types.go]: create a issue for this: https://github.com/envoyproxy/gateway/issues/1786

Comment (tmsnan) [reply api/config/v1alpha1/metric_types.go]: > @AliceProxy @kflynn and I discussed this in the meeting today are here are some thoughts
> 
> * lets create a separate issue for Histogram buckets and better understand the type of buckets users wants to configure
> * the current API is not very intuitive
> 
> ```
> metrics:
>   customMetricScale:
>     statsMatcher:
>       inclusionMatches:
>       - type:  ...
>         value: ...
> ```
> 
> there seem be multiple layers of `match` which feels cumbersome
> 
> Instead can we stick to the earlier suggestion
> 
> ```
> metrics:
>   matches:
>   - type: ...
>      value: ...
> ```
> 
> and lets document which metrics/stats are computed when
> 
> * when `matches` is `nil`
> * when `matches` is not empty

done,using `matches` directly seems convenient to use.

- When Matches is nil(not set), gateway by default create and expose only a subset of Envoy stats with prefixes(cluster_manager,listener_manager,server,cluster.xds-grpc) expressions match on the name of stats.

- When Matches is not empty, this option is to control creation of additional Envoy stats with prefix, suffix, and regex expressions match on the name of the stats.

File: api/config/v1alpha1/metric_types.go [hunk 3]
```diff
@@ -10,6 +10,14 @@ type ProxyMetrics struct {
 	Prometheus *PrometheusProvider `json:"prometheus,omitempty"`
 	// Sinks defines the metric sinks where metrics are sent to.
 	Sinks []MetricSink `json:"sinks,omitempty"`
+	// Matches defines configuration for metric matching rules
```
Comment (arkodg) [inline api/config/v1alpha1/metric_types.go:13]:
```suggestion
	// Matches defines configuration for selecting specific metrics instead of generating all metrics stats that are enabled by default. This helps reduce CPU and memory overhead in Envoy.
```

File: api/config/v1alpha1/metric_types.go [hunk 4]
```diff
@@ -29,6 +37,19 @@ type MetricSink struct {
 	OpenTelemetry *OpenTelemetrySink `json:"openTelemetry,omitempty"`
 }
 
+type Match struct {
```
Comment (arkodg) [inline api/config/v1alpha1/metric_types.go:40]: // Match defines the stats match configuration.

File: api/config/v1alpha1/metric_types.go [hunk 5]
```diff
@@ -29,6 +37,19 @@ type MetricSink struct {
...
+	Type  MatcherType `json:"type"`
+	Value string      `json:"value"`
+}
+
```
Comment (arkodg) [inline api/config/v1alpha1/metric_types.go:44]: // MatcherType defines the stats matcher type
// 
// +kubebuilder:validation:Enum=RegularExpression;Prefix;Suffix

File: api/config/v1alpha1/metric_types.go [hunk 6]
```diff
@@ -29,6 +37,19 @@ type MetricSink struct {
...
+type MatcherType string
+
+const (
+	Prefix MatcherType = "Prefix"
+	Regex  MatcherType = "Regex"
```
Comment (arkodg) [inline api/config/v1alpha1/metric_types.go:49]: lets call this `RegularExpression` keeping it inline with upstream https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HeaderMatchType

**Comment by arkodg** [comment]
thanks for incorporating the suggestions @tmsnan, had some more minor suggestions around comments and doc string, overall LGTM

File: api/config/v1alpha1/metric_types.go [hunk 7]
```diff
@@ -10,6 +10,14 @@ type ProxyMetrics struct {
 	Prometheus *PrometheusProvider `json:"prometheus,omitempty"`
 	// Sinks defines the metric sinks where metrics are sent to.
 	Sinks []MetricSink `json:"sinks,omitempty"`
+	// Matches defines configuration for selecting specific metrics instead of generating all metrics stats
+	// that are enabled by default. This helps reduce CPU and memory overhead in Envoy.
+	// When Matches is nil(not set), gateway by default create and expose only a subset of
```
Comment (arkodg) [inline api/config/v1alpha1/metric_types.go:15]: lines 15-19 can be removed, imo
also because its hard to define in docs what exactly the default stats are, for e.g. we currently also output `http` stats

**Review by zirain — APPROVED** [review]

**Review by arkodg — APPROVED** [review]
LGTM thanks !
